### ISSUE TITLE
Add cron type to job description

### ISSUE THREAD

**Issue by robzienert** [hashicorp/nomad#144]
We have the ability to create "service" and "batch" jobs, but it'd be really nice to have an additional "cron" type. In this type, a task would be created whenever the cron triggers and once the task is complete (be it the process exits gracefully or not), the task is not recreated until the next cron trigger.

The type would not allocate a new task if the number of tasks are already at the maximum count for the group. This might be a nice configuration item to control how this works; performing a no-op, replacement or forced addition.

Assuming we think this a good idea, I'd be happy to jump into it.

**Comment by cbednarski** [comment]
Thanks for the suggestion! I think we will need to start with a design doc to iron out some of the expected behavior / use cases.

For example, "run this on every machine on my cluster once a day" or "run this every 5 minutes (I don't care which machine it runs on)." or "don't overlap if a job is already running".

**Comment by robzienert** [comment]
I think you hit the nail on the head with the 3 examples you gave. Expanding on that, the cron expression would make sense to me to adhere to

**Comment by robzienert** [comment]
Oops.

**Comment by jippi** [comment]
:+1:

**Comment by stevenpall** [comment]
+1

**Comment by ketzacoatl** [comment]
I have also considered Nomad as a simple distributed solution to running various server/cluster cleanup jobs through cron - I would love the ability to schedule a task to run on specific nodes (use tags, or worker capabilities?), all nodes, and at various schedules (X time of day, every X period, etc).

**Comment by c4milo** [comment]
I would start by looking at how https://github.com/mesos/chronos does it.

**Comment by c4milo** [comment]
@armon, @cbednarski I'm in need of this feature, is blocking me right now to get something done. So, I am available to work on this :dancer:. Do you have any update @robzienert?

**Comment by c4milo** [comment]
How about a service outside of Nomad, with an HTTP API, that receives nomad's job configurations as well as its repeating interval (ISO8601). It stores this information in a Consul K/V store and polls it at a give interval using Check-and-Set operations to avoid double scheduling when running multiple instances. It then calls Nomad every time a job needs to run using the Batch scheduler. It could also eventually provide a UI and support job dependencies ala Chronos. 

Does this sound reasonable?

**Comment by ryanuber** [comment]
@c4milo this is something we want to support first-class in Nomad. Our current thinking is that this will be part of Nomad 0.3. If you've got some ideas on how this should work, it would be good to get those into a Google doc to collaborate on the design and make sure everyone is in alignment.

**Comment by c4milo** [comment]
@ryanuber, first-class support is even better. If anybody from Hashicorp is currently working on this I can switch to other pressing tasks I have. Otherwise, I could spend time taking a stab to sending a PR.

**Comment by Sdedelbrock** [comment]
with the v.2 release (congrats btw), this is the only big feature we are waiting on.  We are also willing to contribute to the specs, as well as implementation.

**Comment by c4milo** [comment]
From of the top of my head this is what I may expect ( be aware that I'm still getting familiar with Nomad source code)

* Ability to specify which of the tasks in a task group use cron execution. This is mainly because in my specific use case I may need to co-locate them along with other service tasks.
* Ability to pause cron executed tasks
* Ability to specify a webhook URL to let a third-party service know the exit status of the task.
* Ability to get the history record of executions and exit statuses: last 5 runs? If more than that is possible, great. 
* Ability to retry failed executions if indicated to do so
* CLI option for one-off executions of a given cron task
* Support for ISO8601 Repeatable interval notation
* Document expected behavior in case of network partitions

**Comment by marceldegraaf** [comment]
@robzienert, @c4milo: any progress on this?

**Comment by c4milo** [comment]
@marceldegraaf it is being worked on by @dadgar here https://github.com/hashicorp/nomad/pull/540

**Comment by marceldegraaf** [comment]
Awesome, thanks, I'll keep an eye on that.

**Comment by dadgar** [comment]
#540 introduces this

**Comment by ketzacoatl** [comment]
Should we open new tickets to request/track the workflow/features described in the original post?

**Comment by dadgar** [comment]
Yeah sure! Just keep in mind until 0.3 the feature isn't "complete". For example, force running a periodic job has yet to be implemented but will be coming. So bear that in mind.
